### PR TITLE
Add libbpf bpf_prog_bind_map() API

### LINKED ISSUE #379: Need support to associate manually created map to an eBPF program.
We need implementation for this libbpf API bpf_prog_bind_map that can be used to associate any map to a program.

### PR DESCRIPTION
**Pull request by dthaler**
Also fixes a build dependency from bpftool to force ebpfapi to be built first

Fixes #379

Signed-off-by: Dave Thaler <dthaler@microsoft.com>

### CODE REVIEW

File: libs/platform/kernel/ebpf_platform_kernel.c [hunk 1]
```diff
@@ -43,6 +43,18 @@ __drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_maybenull_
...
+__drv_allocatesMem(Mem) _Must_inspect_result_ _Ret_maybenull_
+    _Post_writable_byte_size_(size) void* ebpf_reallocate(_In_ void* memory, size_t old_size, size_t new_size)
+{
+    void* p = ebpf_allocate(new_size);
+    if (p) {
+        memcpy(p, memory, min(old_size, new_size));
+        if (new_size > old_size)
+            memset(((char*)p) + old_size, 0, new_size - old_size);
+    }
+    return p;
```
Comment (Alan-Jowett) [inline libs/platform/kernel/ebpf_platform_kernel.c:55]: I don't understand. Aren't we leaking the old memory?

Comment (dthaler) [reply libs/platform/kernel/ebpf_platform_kernel.c:56]: Fixed, not sure how that line got lost but now it's back :)

**Review by Alan-Jowett — APPROVED** [review]

File: include/bpf/bpf.h [hunk 2]
```diff
@@ -248,6 +248,18 @@ bpf_obj_pin(int fd, const char* pathname);
  * @{
  */
 
+/**
+ * @brief Bind a map to a program so that it holds a reference on the map.
+ *
+ * @param[in] prog_fd File descriptor of program to bind map to.
+ * @param[in] map_fd File descriptor of map to bind.
+ *
```
Comment (shankarseal) [inline include/bpf/bpf.h:256]: missing bind_opts.

File: libs/execution_context/ebpf_program.c [hunk 3]
```diff
@@ -397,6 +397,30 @@ ebpf_program_type(_In_ const ebpf_program_t* program)
...
+    uint32_t map_count = program->count_of_maps + 1;
+    ebpf_map_t** program_maps =
+        ebpf_reallocate(program->maps, program->count_of_maps * sizeof(ebpf_map_t*), map_count * sizeof(ebpf_map_t*));
+    if (program_maps == NULL) {
+        return EBPF_NO_MEMORY;
+    }
+
+    ebpf_object_acquire_reference((ebpf_object_t*)map);
+    program_maps[map_count - 1] = map;
+    program->maps = program_maps;
```
Comment (shankarseal) [inline libs/execution_context/ebpf_program.c:418]: Does this need synchronization if multiple threads invoke this API?

File: libs/execution_context/ebpf_program.c [hunk 4]
```diff
@@ -397,6 +397,30 @@ ebpf_program_type(_In_ const ebpf_program_t* program)
     return &ebpf_program_get_parameters(program)->program_type;
 }
 
+ebpf_result_t
+ebpf_program_associate_map(ebpf_program_t* program, ebpf_map_t* map)
```
Comment (shankarseal) [inline libs/execution_context/ebpf_program.c:401]: suggest: there is an ebp_program_associate_maps (plural) already. So can the name be changed to reduce confusion and error in future? How about using _single_map  or _additional_map ?

File: include/bpf/bpf.h (hunk 2 above)
Comment (dthaler) [reply include/bpf/bpf.h:278]: Done

File: libs/execution_context/ebpf_program.c (hunk 4 above)
Comment (dthaler) [reply libs/execution_context/ebpf_program.c]: Done

File: libs/execution_context/ebpf_program.c (hunk 3 above)
Comment (dthaler) [reply libs/execution_context/ebpf_program.c:422]: Yes, updated.

**Review by shankarseal — APPROVED** [review]

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**